paz/backend/boxes.py

```diff
@@ -4,377 +4,552 @@
+def process_pair_of_points(image, points, boxes, color, radius):
+    """Process two points to form a box and draw it."""
+    point_A, point_B = take_last_two_points(points)
+    box = order_xyxy(point_A, point_B)
+    paz.draw.box(image, box, color, radius)
+    boxes.append(box)
 
 
-def from_selection(image, radius=5, color=(255, 0, 0), window_name="image"):
-    points, boxes = [], []
-    image = image.copy()
+def on_double_click_event(event, x, y, flags, param, image, points, boxes, radius, color):
+    """Handle double-click event to draw points and boxes."""
+    if event == cv2.EVENT_LBUTTONDBLCLK:
+        draw_circle_and_store_point(image, points, x, y, radius, color)
+        if len(points) % 2 == 0:
+            process_pair_of_points(image, points, boxes, color, radius)
 
-    def order_xyxy(point_A, point_B):
-        (x1, y1) = point_A
-        (x2, y2) = point_B
-        return min(x1, x2), min(y1, y2), max(x1, x2), max(y1, y2)
 
-    def take_last_two_points(points):
-        point_A = points[-1]
-        point_B = points[-2]
-        return point_A, point_B
+def create_on_double_click(image, points, boxes, radius, color):
+    """Create a mouse callback function for double clicks."""
 
     def on_double_click(event, x, y, flags, param):
-        if event == cv2.EVENT_LBUTTONDBLCLK:
-            paz.draw.circle(image, (x, y), radius, color)
-            points.append((x, y))
-            if len(points) % 2 == 0:
-                point_A, point_B = take_last_two_points(points)
-                box = order_xyxy(point_A, point_B)
-                paz.draw.box(image, box, color, radius)
-                boxes.append(box)
-
-    def key_is_pressed(key="q", time=20):
-        return cv2.waitKey(time) & 0xFF == ord(key)
+        on_double_click_event(event, x, y, flags, param, image, points, boxes, radius, color)
+
+    return on_double_click
 
+
+def setup_window_and_callback(image, points, boxes, radius, color, window_name):
+    """Set up the window and mouse callback for box selection."""
+    on_double_click = create_on_double_click(image, points, boxes, radius, color)
     cv2.namedWindow(window_name)
     cv2.setMouseCallback(window_name, on_double_click)
+
+
+def key_is_pressed(key="q", time=20):
+    """Check if a specific key is pressed."""
+    return cv2.waitKey(time) & 0xFF == ord(key)
+
+
+def run_event_loop(image, window_name):
+    """Run the main event loop for displaying the image and handling key presses.
+
+    Args:
+        image (array): The image to display
+        window_name (str): Name of the display window
+
+    Returns:
+        None
+    """
     while True:
         paz.image.show(image, window_name, False)
         if key_is_pressed("q"):
             break
     cv2.destroyWindow(window_name)
-    return jp.array(boxes)
+
+
+def convert_to_center_form(matched_boxes):
+    """Convert matched boxes to center form."""
+    return to_center_form(matched_boxes[:, :4])
```

Comment:
I think the outer function does not convey much more meaning that the inner function. Most likely we should call only the inner function when needed.